### PR TITLE
Add param store for configuration

### PR DESCRIPTION
**Pull request by Calum-Campbell**
- Implement Parameter store for holding configuration
- Use Configraun library to access parameter store within code https://github.com/guardian/configraun 
- Use CLI to get the Play application Secret in the cloudformation userData (we need this on startup)
- Move non secret configuration to CODE.conf/ PROD.conf files.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -18,6 +18,7 @@ libraryDependencies ++= Seq(
   "com.typesafe.akka"       %%  "akka-agent"                % "2.5.4",
   "com.typesafe.akka"       %%  "akka-slf4j"                % "2.5.4",
   "com.amazonaws"           %   "aws-java-sdk-sns"          % "1.10.30",
+  "com.gu"                  %%  "configraun"                % "0.3",
```
Comment (rahilb) [inline build.sbt]: Do we have to use this?

Comment (Calum-Campbell) [reply build.sbt]: It's a library made within the Guardian and  is just a thing wrapper to make accessing parameter store values easier. I don't see why not?

File: app/config/Conf.scala [hunk 2]
```diff
@@ -26,15 +30,55 @@ trait Config {
   val gaCode: String
 }
 
-@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config {
+@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config with Logging {
+
+  implicit val awsClient: AWSSimpleSystemsManagement = AWSSimpleSystemsManagementFactory("eu-west-1", "identity")
+
+  val configraun: com.gu.configraun.models.Configuration = {
```
Comment (jfsoul) [inline app/config/Conf.scala]: I would rather rename 2 `Configuration` imports than use the full classpath.  e.g.

```
import com.gu.configraun.models.{Configuration => ConfigraunConfiguration}
import play.api.{Configuration => PlayConfiguration}
```

File: app/config/Conf.scala [hunk 3]
```diff
@@ -26,15 +30,55 @@ trait Config {
   val gaCode: String
 }
 
-@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config {
+@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config with Logging {
+
+  implicit val awsClient: AWSSimpleSystemsManagement = AWSSimpleSystemsManagementFactory("eu-west-1", "identity")
```
Comment (jfsoul) [inline app/config/Conf.scala]: the string "eu-west-1" is probably a constant in the AWS sdk somewhere e.g. `Regions.EuWest1`

File: app/config/Conf.scala [hunk 4]
```diff
@@ -26,15 +30,55 @@ trait Config {
...
+  val configraun: com.gu.configraun.models.Configuration = {
+    configuration.getOptional[String]("identity-admin.stage") match {
+      case Some("DEV") =>
+        val stack = "identity"
+        val app = "identity-admin"
+        val stage = "DEV"
+        Configraun.loadConfig(Identifier(Stack(stack), App(app), Stage.fromString(stage).get)) match {
+          case Left(a) =>
+            logger.error(s"Unable to load Configraun configuration from AWS (${a.message})")
+            sys.exit(1)
```
Comment (jfsoul) [inline app/config/Conf.scala]: Shouldn't we be throwing an exception rather that using `sys.exit`?

File: app/config/Conf.scala [hunk 5]
```diff
@@ -26,15 +30,55 @@ trait Config {
...
-@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config {
+@Singleton class MainConfig @Inject() (configuration: Configuration) extends Config with Logging {
+
+  implicit val awsClient: AWSSimpleSystemsManagement = AWSSimpleSystemsManagementFactory("eu-west-1", "identity")
+
+  val configraun: com.gu.configraun.models.Configuration = {
+    configuration.getOptional[String]("identity-admin.stage") match {
+      case Some("DEV") =>
+        val stack = "identity"
+        val app = "identity-admin"
```
Comment (jfsoul) [inline app/config/Conf.scala]: I'm not sure I would both declaring these, just put them in the `loadConfig` line

File: app/config/Conf.scala [hunk 6]
```diff
@@ -26,15 +30,55 @@ trait Config {
...
+        val stage = "DEV"
```
Comment (jfsoul) [inline app/config/Conf.scala]: you already have this string in the pattern match if you wanted to re-use that - although probably better using `Stage.Dev`?

File: cloudformation/identity-admin.yaml [hunk 7]
```diff
@@ -268,7 +276,7 @@ Resources:
 
             # Download, extract and chown
             aws s3 cp s3://identity-artifacts/${Stage}/${App}/${App}.zip /${App}
-            aws s3 cp s3://identity-private-config/${Stage}/${App}/${App}.conf /etc/gu/${App}.conf
+            aws ssm get-parameters --name "/${Stack}/${App}/${Stage}/playSecret" --with-decryption --region eu-west-1 | jq -r '.Parameters[] | .Value  | "play.http.secret.key=\"\(.)\""' > /etc/gu/${App}.conf
```
Comment (jfsoul) [inline cloudformation/identity-admin.yaml]: you could substitute region in here instead of hardcoding it

Comment (jfsoul) [reply cloudformation/identity-admin.yaml]: how come not using `get-parameter` as this is a single parameter?

Comment (jfsoul) [reply cloudformation/identity-admin.yaml]: in terms of jq usage - you can use the aws cli global options `--output` and `--query` to filter and format the response instead which I think is more idiomatic.

File: app/config/Conf.scala (hunk 4 above)
Comment (Calum-Campbell) [reply app/config/Conf.scala]: Spoke to Luke about this and this was suggested as you want the application to stop if the correct configuration is not in place. It is implemented in the same way in security HQ. 
https://github.com/guardian/security-hq/blob/4b72f41b244bcb256f9e888b2f9b33287af55c4e/hq/app/AppComponents.scala#L44

**Pull request merged**